feat: authz rules grant specific (#3)
vitwit/cosmos-sdk: x/auth/ante/authz_rules_spec.md

# Description

Problem: 

> The existing authorization (authz) module lacks the flexibility to grant permissions (authz grants) for various types of messages along with specific conditions or rules. This limitation constrains users from customizing their transaction behavior based on specific needs or strategies.

Reviewer: ```suggestion
The current authorization (authz) module lacks the flexibility needed to grant permissions (authz grants) for various types of messages along with specific conditions or rules. This limitation prevents users from customizing their transaction behavior according to specific needs or strategies.

To address this issue, we propose enhancing the authz module to support more granular permissions and conditional rules, allowing for greater customization and control over transaction authorization.
```

## Specific Examples of Limitations:

> Swapping Reward Tokens:
>  - Currently, users cannot set a rule to swap their reward tokens or any other tokens for another token with a specified limit.

Reviewer: ```suggestion
Managing Reward Tokens:
At present, users are able to restake their tokens via authz. But it can do more. Currently users are unable to establish rules for swapping their reward tokens as a strategy as it requires IBCTransfer or PacketForward msgs access. It's not secure to give this grant currently as the recipient address can be anything and grantee can behave maliciously. But if there's a way to restrict recipient address to match with granter's address, this problem is solved. This functionality is necessary to enable users to automate and customize their token management strategies effectively.
```

Sending Tokens to Selected Addresses:

>  - Users are unable to authorize sending tokens to a pre-defined or selected address, restricting the ability to control where tokens are transferred.

Reviewer: ```suggestion
 - Users currently cannot authorize sending tokens to a pre-defined or selected address. This restriction limits their ability to control and automate the transfer of tokens to specific recipients, thereby reducing the efficiency and flexibility of their token management strategies. For example, if an organization wants to authorize an accountant to process salaries every month, the current system's limitations prevent this. Implementing an authz grant to recurrently allow a user to send a specified amount to certain accounts would solve this issue. This feature would automate salary payments, ensuring timely and accurate transactions while reducing administrative overhead.
```

> Staking Tokens with Limitations:

Reviewer: replace it with gov usecase..

Reviewer: also, consider spec format of cosmos-sdk. It describes the problem, but clearly explain about solution, conclusion etc instead of PR diff.
